In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [ ]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
# This step centers each feature (mean = 0), and scales each feature (standard deviation = 1). 
# This step sets the data up for K-Means and PCA since they are distance-based algorithms. 
# By applying the .fit_transform directly on the DataFrame, it is possible to change the DataFrame into a NumPy array. 
scaler = StandardScaler()
market_data_scaled = scaler.fit_transform(df_market_data)


In [6]:
# Create a DataFrame with the scaled data
# Since market_data_scaled is a NumPy array, it has no column names. The following step creates a pandas DataFrame from the NumPy array.
# I reassigned the original columns so that the scaled data keeps the same column names. 
# I also reassigned the original index to the new DataFRame for consistency. 
scaled_market_data_df = pd.DataFrame(market_data_scaled, columns=df_market_data.columns)

# Copy the crypto names from the original DataFrame
crypto_names = df_market_data.index

# Set the coin_id column as index
scaled_market_data_df.index = crypto_names

# Display the scaled DataFrame
scaled_market_data_df.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [ ]:
# Create a list with the number of k-values from 1 to 11
# This step essentially creates a list of numbers, 1 - 10, that represent all the possible number of clusters to be tested when running KMeans. 
k_values = list(range(1,11))

In [8]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for k in k_values:
    kmeans_model = KMeans(n_clusters=k, random_state=1)
    kmeans_model.fit(scaled_market_data_df)
    inertia.append(kmeans_model.inertia_)


c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:14

In [ ]:
# Create a dictionary with the data to plot the Elbow curve
# The dictionary consists of key value pairs: "k" is the list of k_values and "inretia" is the list of inertia values. 
# This step prepares the data in a structured way so that I can easily create the DataFrame for plotting. 
elbow_data = {"k": k_values, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

In [44]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
# While plotting the elbow_data, I assigned the plot to a variable so that I can later call upon it for contrasting with the other plot.  
original_elbow = elbow_df.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k_values)
original_elbow

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Looking at the above elbow curve, the best `k` value is 4. We can see that the inertia drops sharply from k=1 to k=4. After k4, the inertia has a flatter slpoe. This would mean that the data is well grouped into 4 clusters. Any additinal clusters would not improve the model other than making it more complex. 

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [ ]:
# Initialize the K-Means model using the best value for k
# Based on the finding of the preceding elbow plot, 4 is decided to be the best value for k.
# Consequently, n_clusters=4 tells the model to form 4 clusters. 
# I used the random_state=1 to make the results reproducible. Not setting it could generate slightly different clusterings every time it is run. 
model = KMeans(n_clusters=4, random_state=1)

In [14]:
# Fit the K-Means model using the scaled DataFrame
# By using .fit(), the KMeans model will be trained using the scaled data. 
model.fit(scaled_market_data_df)

c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [15]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
# This step will tell the trained KMeans model to assign each cryptocurrency in the scaled data to one of the four clusters. 
crypto_clusters = model.predict(scaled_market_data_df)

# Print the resulting array of cluster values.
print(crypto_clusters)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [16]:
# Create a copy of the scaled DataFrame
crypto_market_predictions = scaled_market_data_df.copy()

In [34]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
crypto_market_predictions["crypto_cluster"] = crypto_clusters

# Display the copy of the scaled DataFrame
crypto_market_predictions.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  Crypto Cluster  crypto_cluster  
coin_id                                                                      
bitcoin                           -0.251637               2               2  
ethereum                          -0.199352               2               2  
tether                            -0.282061               0               0  
ripple                            -0.295546               0               0  
bitcoin-cash                      -0.270317               2               2  
binancecoin                       -0.225533               2               2  
chainlink                         -0.018284               2               2  
cardano                           -0.155428               2               2  
litecoin                          -0.292351               2               2  
bitcoin-cash-sv                   -0.206029               0               0

In [45]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
original_scatter = crypto_market_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="crypto_cluster",
    hover_cols=["coin_id"],
    title="Clustered Cryptocurrencies by Price Change (24h vs 7d)"
)
original_scatter

:NdOverlay   [crypto_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [ ]:
# Create a PCA model instance and set `n_components=3`.
# This step will reduce the original dataset down to just three principal components. 
# The goal is to apply a dimensionality reduction to compress the information in the data while keeping the most important patterns. 
pca = PCA(n_components=3)

In [20]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
# This step does two things: i learns the patterns in the scaled data and projects the data onto the three new directions set above. 
market_data_pca = pca.fit_transform(scaled_market_data_df)

# View the scaled PCA data
market_data_pca[:10]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292]])

In [21]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
# This step will show how important each PC(principal component) is. 
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is 89.5%. Specifically, PC1 captures about 37.20%, PC2 captures 34.70%, and PC3 captures 17.60%. THis setup would mean that the three components preserve almost all of the important information. 

In [22]:
# In this step, the goal is to turn the PCA results into a clean, labeled DataFrame with PC1, PC2, and PC3, and index it by cryptocurrency name. 
# Create a new DataFrame with the PCA data.
crypto_pca_df = pd.DataFrame(
    market_data_pca,
    columns=["PC1", "PC2", "PC3"]
)

# Copy the crypto names from the original scaled DataFrame
names_of_crypto = scaled_market_data_df.index

# Set the coin_id column as index
crypto_pca_df.index=names_of_crypto

# Display the scaled PCA DataFrame
crypto_pca_df.head(10)

PC1       PC2       PC3
coin_id                                      
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [23]:
# Create a list with the number of k-values from 1 to 11
pca_k_values = list(range(1,11))

In [24]:
# Create an empty list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in pca_k_values:
    pca_kmeans_model = KMeans(n_clusters=i, random_state=1)
    pca_kmeans_model.fit(crypto_pca_df)
    pca_inertia.append(pca_kmeans_model.inertia_)

c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:14

In [25]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow_data = {"k":pca_k_values, "inertia": pca_inertia}

# Create a DataFrame with the data to plot the Elbow curve
pca_elbow_df = pd.DataFrame(pca_elbow_data)

In [42]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
# I assigned the plot to a variable to later call upon it for comparisons. 
pca_elbow = pca_elbow_df.hvplot.line(
    x="k",
    y="inertia",
    title="Reduced Crypto Data Elbow Curve",
    xticks=pca_k_values,
    color="red"
)
pca_elbow

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** As was the case before, the inertia drops sharply from k1 - k4 and the curve flattens out. Therefore, the best `k` after the PCA is still 4. 


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** The `k value` stays the same in both the original and the PCA-reduced data. This suggests that the clusters are strong and stable yet after dimentionality reduction. 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [ ]:
# Initialize the K-Means model using the best value for k
# The best value of k, which is 4 as shown in the elbow plot, is used for the n_clusters. 
pca_kmeans = KMeans(n_clusters=4, random_state=1)

In [31]:
# Fit the K-Means model using the PCA data
pca_kmeans.fit(crypto_pca_df)

c:\Users\sofon\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1419: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [32]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_crypto_clusters = pca_kmeans.predict(crypto_pca_df)

# Print the resulting array of cluster values.
print(pca_crypto_clusters)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [36]:
# Create a copy of the scaled PCA DataFrame
crypto_pca_predictions = crypto_pca_df.copy() 

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
crypto_pca_predictions["pca_crypto_clusters"] = pca_crypto_clusters

# Display the copy of the scaled PCA DataFrame
crypto_pca_predictions.head(10)

PC1       PC2       PC3  pca_crypto_clusters
coin_id                                                           
bitcoin         -0.600667  0.842760  0.461595                    2
ethereum        -0.458261  0.458466  0.952877                    2
tether          -0.433070 -0.168126 -0.641752                    0
ripple          -0.471835 -0.222660 -0.479053                    0
bitcoin-cash    -1.157800  2.041209  1.859715                    2
binancecoin     -0.516534  1.388377  0.804071                    2
chainlink       -0.450711  0.517699  2.846143                    2
cardano         -0.345600  0.729439  1.478013                    2
litecoin        -0.649468  0.432165  0.600303                    2
bitcoin-cash-sv -0.759014 -0.201200 -0.217653                    0

In [43]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
# I assigned the plot to a variable for later comparison.  
pca_scatter = crypto_pca_predictions.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="pca_crypto_clusters",
    hover_cols=["coin_id"],
    title="Cryptocurrencies Clustered by PCA Components (PC1 vs PC2)"
)
pca_scatter

:NdOverlay   [pca_crypto_clusters]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [50]:
# Composite plot to contrast the Elbow curves
# For this step, I simply called upon the two elbow plotsas opposed to re-plotting them again. 
original_elbow + pca_elbow

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [52]:
# Composite plot to contrast the clusters
# This is achieved by calling upon the two scatterplots assigned to variables. 
original_scatter + pca_scatter

:Layout
   .NdOverlay.I  :NdOverlay   [crypto_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [pca_crypto_clusters]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Utilizing a reduced number of principal component features in KMeans resulted in clustering that was more stable and less influenced by noise. It did so by reducing the number of features from many original variables, seven to be exact, down to three components (PC1, PC2, and PC3). As stated earlier on, the clustering captured the main patterns in the data, explaining around 89.5% of the original variance. With the use of fewer features, the clusters also appeared to be more compact and relatively better separated. This indicates that the PCA reduced the noise and redundancy. 

  **In conclusion, reducing the feature space simplified the clustering process, making it more straightforward to interpret, although some details may have been omitted.**  